docs/design/mesos-style.md

```diff
@@ -0,0 +1,162 @@
+1. Admission control, quota, applying defaults, etc. runs on each pod; if there are multiple schedulers, one of the admission
+controllers will write the scheduler name as an annotation based on a policy
+1. If a pod is admitted, it is persisted
+1. The appropriate scheduler assigns a nodeName to the instance, which triggers the Kubelet to start the pod's container(s)
+1. Kubelet starts the container(s)
+1. The controller corresponding to the collection manages the pod and the collection during their lifetime
+
+Because Borg-style abstractions are not application-specific, it is common to end up with "two-level scheduling" when
+using Borg. For example, the MapReduce master might create a Borg Job with a certain number of instances with a particular resource shape,
+and then schedule the individual units of map and reduce work onto those instances itself.
+
+## What is a Mesos-style architecture?
+
+Mesos is fundamentally designed to support multiple application-specific "frameworks." A framework is
+composed of a "framework scheduler" and a "framework executor." We will abbreviate "framework scheduler"
+as "framework" since "scheduler" means something very different in Kubernetes (something that just
+assigns pods to nodes).
+
+Unlike Borg and Kubernetes, where there is a single logical endpoint that receives all API requests (the Borgmaster and API server,
+respectively), in Mesos every framework is a separate API endpoint. Mesos does not have any standard set of
+collection abstractions, controllers/state machines, or schedulers; the logic for all of these things is contained
+in each application-specific framework individually, for example
+[Spark](https://github.com/apache/spark/blob/a63d9edcfb8a714a17492517927aa114dea8fea0/docs/running-on-mesos.md).
+(Note that the notion of application-specific does sometimes blur into the realm of workload-specific,
+for example [Chronos](https://github.com/mesos/chronos) is a generic framework for batch jobs.
+However, regardless of what set of Mesos frameworks you are using, the key properties remain: each
+framework is its own API endpoint with its own client-facing and internal abstractions, state machines, and scheduler).
+
+A Mesos-style system can avoid two-level scheduling by integrating cluster-level resource allocation logic and
+application-level work allocation logic in a single framework.
+
+## Building a Mesos-style framework on Kubernetes
+
+Implementing the Mesos model on Kubernetes boils down to enabling application-specific collection abstractions,
+controllers/state machines, and scheduling. There are just three steps:
+* Use API plugins to create API resources for your new application-specific collection abstraction(s)
+* Implement controllers for the new abstractions (and for managing the lifecycle of the pods the controllers generate)
+* Implement a scheduler with the application-specific scheduling logic
```

Comment:
Suppose k8s schedulers have (at some point) a reusable library that makes tracking cluster state simple so that intelligent pod placement becomes easier for scheduler authors over time. How does this impact the scalability of a k8s cluster, with multiple schedulers constantly tracking the full picture of available cluster resources across all of the nodes? I think this is part of the trade-off @bgrant0607